### ISSUE TITLE
So, what is ambitions?

### ISSUE THREAD

**Issue by colesnicov** [Basheyev/CVM#3]
Do you have any reference to the available functions or as I have methods for working with text? Can I add some custom C methods? I read about it in Habr, but at the moment I'm kind of lost in it :/

**Comment by Basheyev** [comment]
Yes, you can add your custom methods. CVM compiler is pretty simple and actually trivial compiler that supports **only integer type (at the moment)**, despite VM have a syscall 0x20 that can print C-style strings. In order to add your own functions to CVM you need to modify Virtual Machine and also Compiler. 

All built-in functions compiled as VM system calls, so if you want to add new syscall of VM you should modify runtime/VirtualMachine.cpp **void sysCall(uint32_t callNo)** and add new syscall number and syscall implementation. CVM is stack virtual machine, so all arguments of function call pulled from the stack of VM, and return values also pushed to the stack.

//---------------------------------------------------------------------
// SYSCALL implementation
//---------------------------------------------------------------------
`void VirtualMachine::sysCall(WORD n) {
        WORD ptr, a;
        switch (n) {
        case 0x20:  // print C style string
                ptr = memory[sp++];
		std::cout << ((char*)&memory[ptr]);
		return;
	case 0x21:  // print int from Top Of the Stack
		a = memory[sp++];
		std::cout << a << endl;
		return;
	case 0x22:
		std::cin >> a;
		memory[--sp] = a;
		return;
	case 0x23: // your syscall number
		// Your syscall here
                // pull arguments from stack - memory[sp++]
                // do your functions job
                // push return value to stack - memory[-sp]
		return
	}
}
`


In the compiler part you should modify **compiler/SourceParser.cpp** to add to global symbol table declaration of functions names and set arguments count - make it visible from user code  (names to your system calls).

`//---------------------------------------------------------------------------
// Builds abstract syntax tree
//---------------------------------------------------------------------------
void SourceParser::buildSyntaxTree() {
    currentToken = 0;

    // add iput system function to symbols table (write int to std out)
    Token iput = { TokenType::IDENTIFIER, "iput", 4, 0,0 };
    rootSymbolTable.addSymbol(iput, SymbolType::FUNCTION);
    rootSymbolTable.lookupSymbol(iput)->argCount = 1;

    // add iget system function to symbols table (read int from std in)
    Token iget = { TokenType::IDENTIFIER, "iget", 4, 0,0 };
    rootSymbolTable.addSymbol(iget, SymbolType::FUNCTION);
    rootSymbolTable.lookupSymbol(iget)->argCount = 1;
    
    // add your own function declaration here

    root = parseModule(&rootSymbolTable);
}`

And modify compiler/CodeGenerator::emitCall, to generate VM code of syscall when your functions is called by name.

`void CodeGenerator::emitCall(ExecutableImage* img, TreeNode* node) {
    
    // look up function name in symbols table
    Token funcToken = node->getToken();
    Symbol* func = node->getSymbolTable()->lookupSymbol(funcToken);
    if (func == NULL || func->type != SymbolType::FUNCTION) raiseError("Function not found.");

    // emit arguments expressions
    for (int i = 0; i < node->getChildCount(); i++) {
        emitExpression(img, node->getChild(i));
    }
        
    // you can't add your function code generation here

    if (funcToken.length == 4 && strncmp(funcToken.text, "iput", 4) == 0) img->emit(OP_SYSCALL, 0x21); else
    if (funcToken.length == 4 && strncmp(funcToken.text, "iget", 4) == 0) img->emit(OP_SYSCALL, 0x22); 
    else {
        // user function
        WORD funcAddress = func->address;
        img->emit(OP_CALL, funcAddress, (WORD) node->getChildCount());
    }
}`

**With this knowledge you can create your custom function only with integer arguments**, But to work with string you need to add new type and modify compiler. Its tricky, because it needs to solve constant string memory layout, support for arrays and possibly dynamic memory allocation in VM.

that's pretty much of it :)
